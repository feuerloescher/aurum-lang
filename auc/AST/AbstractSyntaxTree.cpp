/**
 * auc - The Aurum compiler
 * Copyright (c) 2016  Max Mertens <max.mail@dameweb.de>
 * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
 */

#include "AbstractSyntaxTree.hpp"

using namespace AST;

AbstractSyntaxTree::AbstractSyntaxTree() : irBuilder(llvmContext) {
    llvmModule = llvm::make_unique<llvm::Module>("llvm module", llvmContext);
}

DeclarationList& AbstractSyntaxTree::getDeclarations() {
    return declarations;
}

ASTMap<FunctionDef*>& AbstractSyntaxTree::getFunctionDefs() {
    return functionDefs;
}

ASTMap<ASTPtr<Type>>& AbstractSyntaxTree::getTypes() {
    return types;
}

llvm::IRBuilder<>& AbstractSyntaxTree::getIRBuilder() {
    return irBuilder;
}

llvm::LLVMContext& AbstractSyntaxTree::getLLVMContext() {
    return llvmContext;
}

llvm::Module& AbstractSyntaxTree::getLLVMModule() {
    return *llvmModule;
}
