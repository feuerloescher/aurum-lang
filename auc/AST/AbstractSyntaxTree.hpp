/**
 * auc - The Aurum compiler
 * Copyright (c) 2016  Max Mertens <max.mail@dameweb.de>
 * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
 */

#ifndef AUC_ABSTRACTSYNTAXTREE_HPP
#define AUC_ABSTRACTSYNTAXTREE_HPP

#include "common.hpp"
#include "DeclarativeExpressions.hpp"
#include "ASTPass.hpp"

namespace AST {

class AbstractSyntaxTree {

protected:
    DeclarativeExprList expressions;

public:
    AbstractSyntaxTree(DeclarativeExprList expressions);

    DeclarativeExprList& getExpressions();

}; // class AbstractSyntaxTree

} // namespace AST

#endif // AUC_ABSTRACTSYNTAXTREE_HPP
