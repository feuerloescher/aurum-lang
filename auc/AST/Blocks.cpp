/**
 * auc - The Aurum compiler
 * Copyright (c) 2016  Max Mertens <max.mail@dameweb.de>
 * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
 */

#include "Blocks.hpp"

using namespace AST;

Block::Block() : parentBlock(nullptr), llvmBlock(nullptr) {
}

void Block::runPass(ASTPass& pass) {
    pass.runOn(*this);
}

StatementList& Block::getStatements() {
    return statements;
}

void Block::push_back(ASTPtr<Statement> statement) {
    statements.push_back(statement);
}

ASTMap<VariableDefStmt*>& Block::getVariables() {
    return variables;
}

Block* Block::getParentBlock() {
    return parentBlock;
}

void Block::setParentBlock(Block* block) {
    parentBlock = block;
}

llvm::BasicBlock* Block::getLLVMBlock() {
    return llvmBlock;
}

void Block::setLLVMBlock(llvm::BasicBlock* llvmBlock) {
    this->llvmBlock = llvmBlock;
}


ASTPtr<Expression> IfStmt::getCondition() {
    return condition;
}

Block& IfStmt::getBody() {
    return body;
}

void IfStmt::runPass(ASTPass& pass) {
    pass.runOn(*this);
}

ASTPtr<Expression> WhileLoop::getCondition() {
    return condition;
}

Block& WhileLoop::getBody() {
    return body;
}

void WhileLoop::runPass(ASTPass& pass) {
    pass.runOn(*this);
}
