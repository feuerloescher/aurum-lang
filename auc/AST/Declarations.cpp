/**
 * auc - The Aurum compiler
 * Copyright (c) 2016  Max Mertens <max.mail@dameweb.de>
 * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
 */

#include "Declarations.hpp"

using namespace AST;

FunctionDef::FunctionDef(std::string name, ASTPtr<TypeStmt> type)
    : name(name), type(type) {
}

void FunctionDef::runPass(ASTPass& pass) {
    pass.runOn(*this);
}

std::string FunctionDef::getName() {
    return name;
}

ASTPtr<TypeStmt> FunctionDef::getTypeStmt() {
    return type;
}

VariableDefStmtList& FunctionDef::getParameters() {
    return parameters;
}

Block& FunctionDef::getBody() {
    return body;
}

llvm::Function* FunctionDef::getLLVMFunction() {
    return llvmFunction;
}

void FunctionDef::setLLVMFunction(llvm::Function* llvmFunction) {
    this->llvmFunction = llvmFunction;
}
