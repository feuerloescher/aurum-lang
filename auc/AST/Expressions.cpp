/**
 * auc - The Aurum compiler
 * Copyright (c) 2016  Max Mertens <max.mail@dameweb.de>
 * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
 */

#include "Expressions.hpp"
#include "ASTPass.hpp"
#include "Type.hpp"
#include "TypeStmt.hpp"

using namespace AST;

Expression::Expression(Block* parentBlock, CodeLocation codeLocation)
    : Statement(parentBlock, codeLocation), llvmValue(nullptr) {
}

ASTPtr<Type> Expression::getType() {
    return type;
}

void Expression::setType(ASTPtr<Type> type) {
    this->type = type;
}

llvm::Value* Expression::getLLVMValue() {
    return llvmValue;
}

void Expression::setLLVMValue(llvm::Value* llvmValue) {
    this->llvmValue = llvmValue;
}


FunctionCallExpr::FunctionCallExpr(std::string name, Block* parentBlock,
    CodeLocation codeLocation)
    : Expression(parentBlock, codeLocation), name(name), functionDef(nullptr) {
}

void FunctionCallExpr::runPass(ASTPass& pass) {
    pass.runOn(*this);
}

std::string FunctionCallExpr::getName() {
    return name;
}

ASTList<Expression>& FunctionCallExpr::getArgs() {
    return args;
}

FunctionDef* FunctionCallExpr::getFunctionDef() {
    return functionDef;
}

void FunctionCallExpr::setFunctionDef(FunctionDef* functionDef) {
    this->functionDef = functionDef;
}


MethodCallExpr::MethodCallExpr(ASTPtr<Expression> objectExpr, std::string name,
    Block* parentBlock, CodeLocation codeLocation)
    : Expression(parentBlock, codeLocation), objectExpr(objectExpr), name(name),
    methodDef(nullptr) {
}

void MethodCallExpr::runPass(ASTPass& pass) {
    pass.runOn(*this);
}

ASTPtr<Expression> MethodCallExpr::getObjectExpr() {
    return objectExpr;
}

std::string MethodCallExpr::getName() {
    return name;
}

std::string MethodCallExpr::getMangledName() {
    return objectExpr->getType()->getName() + '.' + name;
}

ASTList<Expression>& MethodCallExpr::getArgs() {
    return args;
}

MethodDef* MethodCallExpr::getMethodDef() {
    return methodDef;
}

void MethodCallExpr::setMethodDef(MethodDef* methodDef) {
    this->methodDef = methodDef;
}

bool MethodCallExpr::isOperator() {
    char first = this->name.at(0);
    return (this->args.size() > 1) ||
        ! (first == '_' || (first >= 'A' && first <= 'Z')
        || (first >= 'a' && first <= 'z'));
}


ConstIntExpr::ConstIntExpr(std::string valueStr, uint8_t numValue,
    Block* parentBlock, CodeLocation codeLocation)
    : Expression(parentBlock, codeLocation),
    valueStr(valueStr), numValue((uint64_t) numValue),
    typeStmt(std::make_shared<TypeStmt>("uint8", CodeLocation::none)) {
}
ConstIntExpr::ConstIntExpr(std::string valueStr, int8_t numValue,
    Block* parentBlock, CodeLocation codeLocation)
    : Expression(parentBlock, codeLocation),
    valueStr(valueStr), numValue((uint64_t) numValue),
    typeStmt(std::make_shared<TypeStmt>("int8", CodeLocation::none)) {
}
ConstIntExpr::ConstIntExpr(std::string valueStr, uint16_t numValue,
    Block* parentBlock, CodeLocation codeLocation)
    : Expression(parentBlock, codeLocation),
    valueStr(valueStr), numValue((uint64_t) numValue),
    typeStmt(std::make_shared<TypeStmt>("uint16", CodeLocation::none)) {
}
ConstIntExpr::ConstIntExpr(std::string valueStr, int16_t numValue,
    Block* parentBlock, CodeLocation codeLocation)
    : Expression(parentBlock, codeLocation),
    valueStr(valueStr), numValue((uint64_t) numValue),
    typeStmt(std::make_shared<TypeStmt>("int16", CodeLocation::none)) {
}
ConstIntExpr::ConstIntExpr(std::string valueStr, uint32_t numValue,
    Block* parentBlock, CodeLocation codeLocation)
    : Expression(parentBlock, codeLocation),
    valueStr(valueStr), numValue((uint64_t) numValue),
    typeStmt(std::make_shared<TypeStmt>("uint32", CodeLocation::none)) {
}
ConstIntExpr::ConstIntExpr(std::string valueStr, int32_t numValue,
    Block* parentBlock, CodeLocation codeLocation)
    : Expression(parentBlock, codeLocation),
    valueStr(valueStr), numValue((uint64_t) numValue),
    typeStmt(std::make_shared<TypeStmt>("int32", CodeLocation::none)) {
}
ConstIntExpr::ConstIntExpr(std::string valueStr, uint64_t numValue,
    Block* parentBlock, CodeLocation codeLocation)
    : Expression(parentBlock, codeLocation),
    valueStr(valueStr), numValue((uint64_t) numValue),
    typeStmt(std::make_shared<TypeStmt>("uint64", CodeLocation::none)) {
}
ConstIntExpr::ConstIntExpr(std::string valueStr, int64_t numValue,
    Block* parentBlock, CodeLocation codeLocation)
    : Expression(parentBlock, codeLocation),
    valueStr(valueStr), numValue((uint64_t) numValue),
    typeStmt(std::make_shared<TypeStmt>("int64", CodeLocation::none)) {
}

std::string ConstIntExpr::getValueStr() {
    return valueStr;
}

void ConstIntExpr::runPass(ASTPass& pass) {
    pass.runOn(*this);
}

uint64_t ConstIntExpr::getNumValue() {
    return numValue;
}

ASTPtr<TypeStmt> ConstIntExpr::getTypeStmt() {
    return typeStmt;
}


VariableExpr::VariableExpr(std::string name, Block* parentBlock,
    CodeLocation codeLocation) : Expression(parentBlock, codeLocation),
    name(name), variableDefStmt(nullptr) {
}

void VariableExpr::runPass(ASTPass& pass) {
    pass.runOn(*this);
}

std::string VariableExpr::getName() {
    return name;
}

VariableDefStmt* VariableExpr::getVariableDefStmt() {
    return variableDefStmt;
}

void VariableExpr::setVariableDefStmt(VariableDefStmt* variableDefStmt) {
    this->variableDefStmt = variableDefStmt;
}
