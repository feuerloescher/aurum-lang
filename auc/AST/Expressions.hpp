/**
 * auc - The Aurum compiler
 * Copyright (c) 2016  Max Mertens <max.mail@dameweb.de>
 * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
 */

#ifndef AUC_EXPRESSIONS_HPP
#define AUC_EXPRESSIONS_HPP

#include "common.hpp"
#include "Statements.hpp"
#include "CodeLocation.hpp"

#include <llvm/IR/Value.h>
#include <string>

namespace AST {

class Expression : public Statement {

protected:
    ASTPtr<Type> type;
    llvm::Value* llvmValue;

public:
    Expression(Block* parentBlock, CodeLocation codeLocation);

    virtual void runPass(ASTPass& pass) = 0;
    ASTPtr<Type> getType();
    void setType(ASTPtr<Type> type);
    llvm::Value* getLLVMValue();
    void setLLVMValue(llvm::Value* llvmValue);

}; // class Expression

typedef ASTList<Expression> ExpressionList;


class FunctionCallExpr : public Expression {

protected:
    std::string name;
    ASTList<Expression> args;
    FunctionDef* functionDef;

public:
    FunctionCallExpr(std::string name, Block* parentBlock,
        CodeLocation codeLocation);

    virtual void runPass(ASTPass& pass);
    std::string getName();
    ASTList<Expression>& getArgs();
    FunctionDef* getFunctionDef();
    void setFunctionDef(FunctionDef* functionDef);

}; // class FunctionCallExpr


class MethodCallExpr : public Expression {

protected:
    ASTPtr<Expression> objectExpr;
    std::string name;
    ASTList<Expression> args;
    MethodDef* methodDef;

public:
    MethodCallExpr(ASTPtr<Expression> tmpObjectExpr, std::string name,
        Block* parentBlock, CodeLocation codeLocation);

    virtual void runPass(ASTPass& pass);
    ASTPtr<Expression> getObjectExpr();
    std::string getName();
    std::string getMangledName();
    ASTList<Expression>& getArgs();
    MethodDef* getMethodDef();
    void setMethodDef(MethodDef* functionDef);

}; // class MethodCallExpr


class ConstIntExpr : public Expression {

protected:
    std::string valueStr;
    uint64_t numValue;
    ASTPtr<TypeStmt> typeStmt;

public:
    ConstIntExpr(std::string valueStr, uint8_t numValue,
        Block* parentBlock, CodeLocation codeLocation);
    ConstIntExpr(std::string valueStr, int8_t numValue,
        Block* parentBlock, CodeLocation codeLocation);
    ConstIntExpr(std::string valueStr, uint16_t numValue,
        Block* parentBlock, CodeLocation codeLocation);
    ConstIntExpr(std::string valueStr, int16_t numValue,
        Block* parentBlock, CodeLocation codeLocation);
    ConstIntExpr(std::string valueStr, uint32_t numValue,
        Block* parentBlock, CodeLocation codeLocation);
    ConstIntExpr(std::string valueStr, int32_t numValue,
        Block* parentBlock, CodeLocation codeLocation);
    ConstIntExpr(std::string valueStr, uint64_t numValue,
        Block* parentBlock, CodeLocation codeLocation);
    ConstIntExpr(std::string valueStr, int64_t numValue,
        Block* parentBlock, CodeLocation codeLocation);

    virtual void runPass(ASTPass& pass);

    std::string getValueStr();
    uint64_t getNumValue();
    ASTPtr<TypeStmt> getTypeStmt();

}; // class ConstIntExpr


class VariableExpr : public Expression {

protected:
    std::string name;
    VariableDefStmt* variableDefStmt;

public:
    VariableExpr(std::string name, Block* parentBlock,
        CodeLocation codeLocation);

    virtual void runPass(ASTPass& pass);

    std::string getName();
    VariableDefStmt* getVariableDefStmt();
    void setVariableDefStmt(VariableDefStmt* variableDefStmt);

}; // class VariableExpr


} // namespace AST

#endif // AUC_EXPRESSIONS_HPP
