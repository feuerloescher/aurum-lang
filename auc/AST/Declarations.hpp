/**
 * auc - The Aurum compiler
 * Copyright (c) 2016  Max Mertens <max.mail@dameweb.de>
 * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
 */

#ifndef AUC_DECLARATIONS_HPP
#define AUC_DECLARATIONS_HPP

#include "common.hpp"
#include "Blocks.hpp"
#include "ASTPass.hpp"

#include <llvm/IR/Function.h>
#include <string>

namespace AST {

class Declaration {

public:
    virtual void runPass(ASTPass& pass) = 0;

}; // class Declaration

typedef ASTList<Declaration> DeclarationList;


class FunctionDef : public Declaration {

protected:
    std::string name;
    ASTPtr<TypeStmt> type;
    ASTList<VariableDefStmt> parameters;
    Block body;
    llvm::Function* llvmFunction;

public:
    FunctionDef(std::string name, ASTPtr<TypeStmt> type);

    virtual void runPass(ASTPass& pass);

    std::string getName();
    ASTPtr<TypeStmt> getTypeStmt();
    VariableDefStmtList& getParameters();
    Block& getBody();
    llvm::Function* getLLVMFunction();
    void setLLVMFunction(llvm::Function* llvmFunction);

}; // class FunctionDef

} // namespace AST

#endif // AUC_DECLARATIONS_HPP
