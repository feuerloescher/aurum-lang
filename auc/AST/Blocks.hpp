/**
 * auc - The Aurum compiler
 * Copyright (c) 2016  Max Mertens <max.mail@dameweb.de>
 * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
 */

#ifndef AUC_BLOCKS_HPP
#define AUC_BLOCKS_HPP

#include "common.hpp"
#include "Statements.hpp"
#include "Expressions.hpp"
#include "ASTPass.hpp"

namespace AST {

class Block : public Statement {

protected:
    StatementList body;

public:
    virtual void runPass(ASTPass& pass) = 0;

    StatementList getBody();

}; // class Block


class IfStmt : public Block {

protected:
    ASTPtr<Expression> condition;

public:
    IfStmt(ASTPtr<Expression> condition);
    virtual void runPass(ASTPass& pass);

}; // class IfStmt


class WhileLoop : public Block {

protected:
    ASTPtr<Expression> condition;

public:
    WhileLoop(ASTPtr<Expression> condition);
    virtual void runPass(ASTPass& pass);

}; // class WhileLoop

} // namespace AST

#endif // AUC_BLOCKS_HPP
