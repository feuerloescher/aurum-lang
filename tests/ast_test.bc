; ModuleID = 'llvm module'

define i8 @"uint8.+"(i8* %this, i8 %param) {
entry:
  %thisval = load i8, i8* %this, align 1
  %addtmp = add i8 %thisval, %param
  ret i8 %addtmp
}

define i8 @uint8.-(i8* %this, i8 %param) {
entry:
  %thisval = load i8, i8* %this, align 1
  %subtmp = sub i8 %thisval, %param
  ret i8 %subtmp
}

define i8 @"uint8.*"(i8* %this, i8 %param) {
entry:
  %thisval = load i8, i8* %this, align 1
  %multmp = mul i8 %thisval, %param
  ret i8 %multmp
}

define i8 @"uint8./"(i8* %this, i8 %param) {
entry:
  %thisval = load i8, i8* %this, align 1
  %divtmp = udiv i8 %thisval, %param
  ret i8 %divtmp
}

define i8 @"uint8.="(i8* %this, i8 %param) {
entry:
  store i8 %param, i8* %this, align 1
  ret i8 %param
}

define i8 @"uint8.++"(i8* %this) {
entry:
  %thisval = load i8, i8* %this, align 1
  %inctmp = add i8 %thisval, 1
  store i8 %inctmp, i8* %this, align 1
  ret i8 %inctmp
}

define i8 @uint8.--(i8* %this) {
entry:
  %thisval = load i8, i8* %this, align 1
  %dectmp = add i8 %thisval, -1
  store i8 %dectmp, i8* %this, align 1
  ret i8 %dectmp
}

define i16 @"uint16.+"(i16* %this, i16 %param) {
entry:
  %thisval = load i16, i16* %this, align 2
  %addtmp = add i16 %thisval, %param
  ret i16 %addtmp
}

define i16 @uint16.-(i16* %this, i16 %param) {
entry:
  %thisval = load i16, i16* %this, align 2
  %subtmp = sub i16 %thisval, %param
  ret i16 %subtmp
}

define i16 @"uint16.*"(i16* %this, i16 %param) {
entry:
  %thisval = load i16, i16* %this, align 2
  %multmp = mul i16 %thisval, %param
  ret i16 %multmp
}

define i16 @"uint16./"(i16* %this, i16 %param) {
entry:
  %thisval = load i16, i16* %this, align 2
  %divtmp = udiv i16 %thisval, %param
  ret i16 %divtmp
}

define i16 @"uint16.="(i16* %this, i16 %param) {
entry:
  store i16 %param, i16* %this, align 2
  ret i16 %param
}

define i16 @"uint16.++"(i16* %this) {
entry:
  %thisval = load i16, i16* %this, align 2
  %inctmp = add i16 %thisval, 1
  store i16 %inctmp, i16* %this, align 2
  ret i16 %inctmp
}

define i16 @uint16.--(i16* %this) {
entry:
  %thisval = load i16, i16* %this, align 2
  %dectmp = add i16 %thisval, -1
  store i16 %dectmp, i16* %this, align 2
  ret i16 %dectmp
}

define i32 @"uint32.+"(i32* %this, i32 %param) {
entry:
  %thisval = load i32, i32* %this, align 4
  %addtmp = add i32 %thisval, %param
  ret i32 %addtmp
}

define i32 @uint32.-(i32* %this, i32 %param) {
entry:
  %thisval = load i32, i32* %this, align 4
  %subtmp = sub i32 %thisval, %param
  ret i32 %subtmp
}

define i32 @"uint32.*"(i32* %this, i32 %param) {
entry:
  %thisval = load i32, i32* %this, align 4
  %multmp = mul i32 %thisval, %param
  ret i32 %multmp
}

define i32 @"uint32./"(i32* %this, i32 %param) {
entry:
  %thisval = load i32, i32* %this, align 4
  %divtmp = udiv i32 %thisval, %param
  ret i32 %divtmp
}

define i32 @"uint32.="(i32* %this, i32 %param) {
entry:
  store i32 %param, i32* %this, align 4
  ret i32 %param
}

define i32 @"uint32.++"(i32* %this) {
entry:
  %thisval = load i32, i32* %this, align 4
  %inctmp = add i32 %thisval, 1
  store i32 %inctmp, i32* %this, align 4
  ret i32 %inctmp
}

define i32 @uint32.--(i32* %this) {
entry:
  %thisval = load i32, i32* %this, align 4
  %dectmp = add i32 %thisval, -1
  store i32 %dectmp, i32* %this, align 4
  ret i32 %dectmp
}

define i64 @"uint64.+"(i64* %this, i64 %param) {
entry:
  %thisval = load i64, i64* %this, align 4
  %addtmp = add i64 %thisval, %param
  ret i64 %addtmp
}

define i64 @uint64.-(i64* %this, i64 %param) {
entry:
  %thisval = load i64, i64* %this, align 4
  %subtmp = sub i64 %thisval, %param
  ret i64 %subtmp
}

define i64 @"uint64.*"(i64* %this, i64 %param) {
entry:
  %thisval = load i64, i64* %this, align 4
  %multmp = mul i64 %thisval, %param
  ret i64 %multmp
}

define i64 @"uint64./"(i64* %this, i64 %param) {
entry:
  %thisval = load i64, i64* %this, align 4
  %divtmp = udiv i64 %thisval, %param
  ret i64 %divtmp
}

define i64 @"uint64.="(i64* %this, i64 %param) {
entry:
  store i64 %param, i64* %this, align 4
  ret i64 %param
}

define i64 @"uint64.++"(i64* %this) {
entry:
  %thisval = load i64, i64* %this, align 4
  %inctmp = add i64 %thisval, 1
  store i64 %inctmp, i64* %this, align 4
  ret i64 %inctmp
}

define i64 @uint64.--(i64* %this) {
entry:
  %thisval = load i64, i64* %this, align 4
  %dectmp = add i64 %thisval, -1
  store i64 %dectmp, i64* %this, align 4
  ret i64 %dectmp
}

define i8 @"int8.+"(i8* %this, i8 %param) {
entry:
  %thisval = load i8, i8* %this, align 1
  %addtmp = add i8 %thisval, %param
  ret i8 %addtmp
}

define i8 @int8.-(i8* %this, i8 %param) {
entry:
  %thisval = load i8, i8* %this, align 1
  %subtmp = sub i8 %thisval, %param
  ret i8 %subtmp
}

define i8 @"int8.*"(i8* %this, i8 %param) {
entry:
  %thisval = load i8, i8* %this, align 1
  %multmp = mul i8 %thisval, %param
  ret i8 %multmp
}

define i8 @"int8./"(i8* %this, i8 %param) {
entry:
  %thisval = load i8, i8* %this, align 1
  %divtmp = sdiv i8 %thisval, %param
  ret i8 %divtmp
}

define i8 @"int8.="(i8* %this, i8 %param) {
entry:
  store i8 %param, i8* %this, align 1
  ret i8 %param
}

define i8 @"int8.++"(i8* %this) {
entry:
  %thisval = load i8, i8* %this, align 1
  %inctmp = add i8 %thisval, 1
  store i8 %inctmp, i8* %this, align 1
  ret i8 %inctmp
}

define i8 @int8.--(i8* %this) {
entry:
  %thisval = load i8, i8* %this, align 1
  %dectmp = add i8 %thisval, -1
  store i8 %dectmp, i8* %this, align 1
  ret i8 %dectmp
}

define i16 @"int16.+"(i16* %this, i16 %param) {
entry:
  %thisval = load i16, i16* %this, align 2
  %addtmp = add i16 %thisval, %param
  ret i16 %addtmp
}

define i16 @int16.-(i16* %this, i16 %param) {
entry:
  %thisval = load i16, i16* %this, align 2
  %subtmp = sub i16 %thisval, %param
  ret i16 %subtmp
}

define i16 @"int16.*"(i16* %this, i16 %param) {
entry:
  %thisval = load i16, i16* %this, align 2
  %multmp = mul i16 %thisval, %param
  ret i16 %multmp
}

define i16 @"int16./"(i16* %this, i16 %param) {
entry:
  %thisval = load i16, i16* %this, align 2
  %divtmp = sdiv i16 %thisval, %param
  ret i16 %divtmp
}

define i16 @"int16.="(i16* %this, i16 %param) {
entry:
  store i16 %param, i16* %this, align 2
  ret i16 %param
}

define i16 @"int16.++"(i16* %this) {
entry:
  %thisval = load i16, i16* %this, align 2
  %inctmp = add i16 %thisval, 1
  store i16 %inctmp, i16* %this, align 2
  ret i16 %inctmp
}

define i16 @int16.--(i16* %this) {
entry:
  %thisval = load i16, i16* %this, align 2
  %dectmp = add i16 %thisval, -1
  store i16 %dectmp, i16* %this, align 2
  ret i16 %dectmp
}

define i32 @"int32.+"(i32* %this, i32 %param) {
entry:
  %thisval = load i32, i32* %this, align 4
  %addtmp = add i32 %thisval, %param
  ret i32 %addtmp
}

define i32 @int32.-(i32* %this, i32 %param) {
entry:
  %thisval = load i32, i32* %this, align 4
  %subtmp = sub i32 %thisval, %param
  ret i32 %subtmp
}

define i32 @"int32.*"(i32* %this, i32 %param) {
entry:
  %thisval = load i32, i32* %this, align 4
  %multmp = mul i32 %thisval, %param
  ret i32 %multmp
}

define i32 @"int32./"(i32* %this, i32 %param) {
entry:
  %thisval = load i32, i32* %this, align 4
  %divtmp = sdiv i32 %thisval, %param
  ret i32 %divtmp
}

define i32 @"int32.="(i32* %this, i32 %param) {
entry:
  store i32 %param, i32* %this, align 4
  ret i32 %param
}

define i32 @"int32.++"(i32* %this) {
entry:
  %thisval = load i32, i32* %this, align 4
  %inctmp = add i32 %thisval, 1
  store i32 %inctmp, i32* %this, align 4
  ret i32 %inctmp
}

define i32 @int32.--(i32* %this) {
entry:
  %thisval = load i32, i32* %this, align 4
  %dectmp = add i32 %thisval, -1
  store i32 %dectmp, i32* %this, align 4
  ret i32 %dectmp
}

define i64 @"int64.+"(i64* %this, i64 %param) {
entry:
  %thisval = load i64, i64* %this, align 4
  %addtmp = add i64 %thisval, %param
  ret i64 %addtmp
}

define i64 @int64.-(i64* %this, i64 %param) {
entry:
  %thisval = load i64, i64* %this, align 4
  %subtmp = sub i64 %thisval, %param
  ret i64 %subtmp
}

define i64 @"int64.*"(i64* %this, i64 %param) {
entry:
  %thisval = load i64, i64* %this, align 4
  %multmp = mul i64 %thisval, %param
  ret i64 %multmp
}

define i64 @"int64./"(i64* %this, i64 %param) {
entry:
  %thisval = load i64, i64* %this, align 4
  %divtmp = sdiv i64 %thisval, %param
  ret i64 %divtmp
}

define i64 @"int64.="(i64* %this, i64 %param) {
entry:
  store i64 %param, i64* %this, align 4
  ret i64 %param
}

define i64 @"int64.++"(i64* %this) {
entry:
  %thisval = load i64, i64* %this, align 4
  %inctmp = add i64 %thisval, 1
  store i64 %inctmp, i64* %this, align 4
  ret i64 %inctmp
}

define i64 @int64.--(i64* %this) {
entry:
  %thisval = load i64, i64* %this, align 4
  %dectmp = add i64 %thisval, -1
  store i64 %dectmp, i64* %this, align 4
  ret i64 %dectmp
}

define i32 @foo(i32 %var) {
entry:
  %var1 = alloca i32, align 4
  store i32 %var, i32* %var1, align 4
  %0 = call i32 @"uint32.+"(i32* nonnull %var1, i32 5)
  ret i32 %0
}

define i32 @main() {
entry:
  %0 = call i32 @foo(i32 5)
  %y = alloca i32, align 4
  %1 = call i32 @"uint32.="(i32* nonnull %y, i32 %0)
  %2 = call i32 @"uint32.++"(i32* nonnull %y)
  %y.val = load i32, i32* %y, align 4
  ret i32 %y.val
}
