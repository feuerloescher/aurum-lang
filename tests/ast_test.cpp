/**
 * auc - The Aurum compiler
 * Copyright (c) 2016  Max Mertens <max.mail@dameweb.de>
 * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
 */

#include "AST/AbstractSyntaxTree.hpp"
#include "AST/ScalarTypes.hpp"
#include "AST/Declarations.hpp"
#include "AST/Statements.hpp"
#include "AST/Blocks.hpp"
#include "AST/TypeStmt.hpp"
#include "AST/Expressions.hpp"
#include "Passes/PrintPass.hpp"
#include "Passes/StdLibPass.hpp"
#include "Passes/IdentifierPass.hpp"
#include "Passes/TypePass.hpp"
#include "Passes/LLVMPass.hpp"
#include "Passes/IRExportPass.hpp"

#include <iostream>

using namespace AST;
using namespace Passes;
using namespace std;

int main() {
    AbstractSyntaxTree ast;

    auto variable = make_shared<VariableDefStmt>("var",
        make_shared<TypeStmt>("uint32"));
    auto funcDecl = make_shared<FunctionDef>("foo",
        make_shared<TypeStmt>("uint32"));
    funcDecl->getParameters().push_back(variable);
    ast.getDeclarations().push_back(funcDecl);

    auto addOp = make_shared<MethodCallExpr>(
        make_shared<VariableExpr>("var"), "+");
    addOp->getParameters().push_back(make_shared<ConstUInt32Expr>(5));
    auto ret = make_shared<ReturnStmt>(addOp);
    funcDecl->getBody().push_back(ret);

    auto mainDecl = make_shared<FunctionDef>("main",
        make_shared<TypeStmt>("uint32"));
    ast.getDeclarations().push_back(mainDecl);

    auto funcCall = make_shared<FunctionCallExpr>("foo");
    funcCall->getParameters().push_back(make_shared<ConstUInt32Expr>(5));
    auto defAssignX = make_shared<VariableDefAssignStmt>("x",
        make_shared<TypeStmt>("uint32"), funcCall);
    mainDecl->getBody().push_back(defAssignX);

    auto defY = make_shared<VariableDefStmt>("y",
        make_shared<TypeStmt>("uint32"));
    mainDecl->getBody().push_back(defY);

    auto negateOp = make_shared<MethodCallExpr>(
        make_shared<VariableExpr>("x"), "-");
    auto assignOp = make_shared<MethodCallExpr>(
        make_shared<VariableExpr>("y"), "=");
    assignOp->getParameters().push_back(negateOp);
    mainDecl->getBody().push_back(assignOp);

    auto unaryAssignOp = make_shared<MethodCallExpr>(
        make_shared<VariableExpr>("y"), "++");
    mainDecl->getBody().push_back(unaryAssignOp);

    ASTPtr<VariableExpr> varY3{new VariableExpr("y")};
    auto ret2 = make_shared<ReturnStmt>(make_shared<VariableExpr>("y"));
    mainDecl->getBody().push_back(ret2);

    StdLibPass stdLibPass(ast);
    stdLibPass.run();

    IdentifierPass identifier(ast);
    identifier.run();

    TypePass typePass(ast);
    typePass.run();

    PrintPass printer(ast, std::cout);
    printer.run();

    LLVMPass llvmPass(ast);
    llvmPass.run();

    IRExportPass exportPass(ast);
    exportPass.run();
    return 0;
}
