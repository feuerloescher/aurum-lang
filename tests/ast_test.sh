#!/bin/sh -x

echo "Warning: this test has been temporarily disabled"
#./ast_test.bin a a a a
#[ "$?" -eq "121" ]
