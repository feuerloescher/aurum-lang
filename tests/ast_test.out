// Aurum to C transpiled code
#define uint32 int

uint32 main(uint32 argc) {
  uint32 x = factorial(argc);
  uint32 y;
  (y = x);
  y++;
  return y;
}

uint32 factorial(uint32 var) {
  if ((var == 0)) {
    return 1;
  };
  return (var * factorial((var - 1)));
}

// End of C code
